pxeboot.py

```diff
@@ -0,0 +1,213 @@
+
+
+children = []
+iso_mount_path = "/mnt/marvel_dpu_iso"
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Process ISO file.")
+    parser.add_argument("iso", type=str, help="Mandatory argument of type string for ISO file (make sure the path to this file was mounted when running the pod via -v /host/iso:/container/iso).")
+    parser.add_argument("--dev", type=str, default="eno4", help="Optional argument of type string for device. Default is 'eno4'.")
+
+    args = parser.parse_args()
+    if not os.path.exists(args.iso):
+        print(f"Couldn't read iso file {args.iso}")
+        raise Exception("Invalid path to iso provided")
+
+    return args
+
+def run_process(cmd):
+    p = Process(target=run, args=(cmd,))
+    p.start()
+    return p
+
+
+def wait_any_ping(hn, timeout):
+    print("Waiting for response from ping")
+    begin = time.time()
+    end = begin
+    while end - begin < timeout:
+        for e in hn:
+            if ping(e):
+                return e
+        time.sleep(5)
+        end = time.time()
+    raise Exception(f"No response after {round(end - begin, 2)}s")
+
+
+def ping(hn):
+    ping_cmd = f"timeout 1 ping -4 -c 1 {hn}"
+    return run(ping_cmd).returncode == 0
+
+
+def wait_for_boot():
+    time.sleep(1000)
+    try:
+        candidates = list(f"172.131.100.{x}" for x in range(10, 21))
+        response_ip = wait_any_ping(candidates, 12000)
+        print(f"got response from {response_ip}")
+    except Exception as e:
+        print("Failed to detect IP from Marvell card")
+        raise e
+
+def select_pxe_entry():
+    print("selecting pxe entry")
+    ESC = "\x1b"
+    KEY_DOWN = '\x1b[B'
+    KEY_ENTER = '\r\n'
+
+    run("pkill -9 minicom")
+    print("spawn minicom")
+    child = pexpect.spawn(minicom_cmd("/dev/ttyUSB0"))
+    child.maxread = 10000
+    print("waiting for instructions to access boot menu")
+    child.expect("Press 'B' within 10 seconds for boot menu", 30)
+    time.sleep(1)
+    print("Pressing B to access boot menu")
+    child.send('b')
+    print("waiting for instructions to Boot from Secondary Boot Device")
+    child.expect("2\) Boot from Secondary Boot Device", 10)
+    time.sleep(1)
+    child.send('2')
+    print("waiting to escape to UEFI boot menu")
+    child.expect("Press ESCAPE for boot options", 60)
+    print("Sending escape 5 times")
+    child.send(ESC*5)
+    print("waiting on language option")
+    child.expect("This is the option.*one adjusts to change.*the language for the.*current system", timeout=3)
+    print("pressing down")
+    child.send(KEY_DOWN)
+    time.sleep(1)
+    print("pressing down again")
+    child.send(KEY_DOWN)
+    print("waiting for Boot manager entry")
+    child.expect("This selection will.*take you to the Boot.*Manager", timeout=3)
+    child.send(KEY_ENTER)
+    child.expect("Device Path")
+    retry = 30
+    print(f"Trying up to {retry} times to find pxe boot interface")
+    while retry:
+        child.send(KEY_DOWN)
```

Comment:
We might get this to interface faster from going up instead of down.

Reply:
True, but I am relying on the page reaching "UEFI PXEv4.*MAC:80AA9988776A" to know that the cursor is on this option (paging down means the most recent option you see is the one the cursor is on)